### DISCUSSION TITLE
Rethinking mapsets

### DISCUSSION THREAD

**Discussion by LastExceed** [ppy/osu#16559] [Ideas]
I'd like to suggest the removal of "mapsets". Instead, beatmaps (or more precisely, the data that is currently stored in .osu files in stable) should each be standalone and their audio resources (music and hitsounds) should be stored separately so that multiple beatmaps can access the same resources.

Especially for songs from featured artists I often find myself spam-clicking through osu!direct downloading the same mp3 40 times just to find the beatmap that I enjoy the most to add to my collection, and then either manually delete all the others just minutes after download or have them clutter my local database (It's not rare for people to have a 50GB+ osu! folder because of this, which imo is insane).

Storing beatmap resources separately would allow even beatmaps from different creators to use the same resources which in turn could decrease the size of the beatmap database a lot (both server and client side). When a DMCA takedown request occurs we'd no longer have to remove the entire mapset but only the mp3, giving the user the option to supply the missing file locally (integrity could be verified via checksum).
The upload limit could be changed to "only X new resources per user" (increased with supporter tag) and no limit (or just a very high limit for sanity) for beatmaps that solely use existing resources as those would consume virtually no new space. This would encourage users to take advantage of the new system.

Ranking criteria could be changed to a per-beatmap basis rather than per-mapset basis. One big problem I see with the current ranking process is the fact that when you want to rank a difficult beatmap of a short song you're forced to create lots of difficulties in order to meet the ranking criteria, which often leads to low quality maps, the urge to get lots of guest diffs, or sometimes discourages mappers from attempting to get ranked entirely. (Thats why you see so many hard maps in the loved section, at least in mania). A per-beatmap ranking process would encourage mappers to put more effort in a single beatmap, leading to higher quality maps and a decrease in low quality maps.

**Comment by Cynplytholowazy** [comment]
How do you deal with currently existing mapsets? There are lots of mapsets that have no metadata whatsoever to determine to used the shared resources. Besides file duplicates are rare among mapsets, considering there are about a million mapsets right now sitting in osu! server + even if they are the same song they might have quality difference.

**Comment by Aergwyn** [comment]
On the level of files themselves this is already done as long as it's actually the same file (matching hash). There shouldn't be duplicate files anymore.

**Comment by peppy** [comment]
We are already moving in this direction. It's going to take much more time before we are in a state this can be realised though.

**Comment by lunafreya** [comment]
And by _only X new resources per user_, you mean that users might have to wait for a little while before uploading new resources - some sort of "_new resource queue_", perhaps?
